public/app/features/alerting/unified/components/rule-editor/labels/LabelsField.tsx

```diff
@@ -0,0 +1,477 @@
+    onClose();
+  };
+  // default values for the subform are the initial labels
+  const defaultValues: LabelsSubformValues = useMemo(() => {
+    return { labelsInSubform: initialLabels };
+  }, [initialLabels]);
+  const formAPI = useForm<LabelsSubformValues>({ defaultValues });
+  return (
+    <FormProvider {...formAPI}>
+      <form onSubmit={formAPI.handleSubmit(onSave)}>
+        <Stack direction="column" gap={4}>
+          <Text>Add labels to your rule for searching, silencing, or routing to a notification policy.</Text>
+          <Stack direction="column" gap={1}>
+            <LabelsWithSuggestions dataSourceName={dataSourceName} />
+            <Space v={2} />
+            <LabelsInRule labels={formAPI.watch('labelsInSubform')} />
+            <Space v={1} />
+            <div className={styles.confirmButton}>
+              <Button type="button" variant="secondary" onClick={onClose}>
+                Cancel
+              </Button>
+              <Button type="submit">Save</Button>
+            </div>
+          </Stack>
+        </Stack>
+      </form>
+    </FormProvider>
+  );
+}
+export function useCombinedLabels(
+  dataSourceName: string,
+  labelsPluginInstalled: boolean,
+  loadingLabelsPlugin: boolean,
+  labelsInSubform: Array<{ key: string; value: string }>
+) {
+  // ------- Get labels keys and their values from existing alerts
+  const { loading, labelsByKey: labelsByKeyFromExisingAlerts } = useGetCustomLabels(dataSourceName);
+  // ------- Get only the keys from the ops labels, as we will fetch the values for the keys once the key is selected.
+  const { loading: isLoadingLabels, labelsOpsKeys = [] } = useGetOpsLabelsKeys(
+    !labelsPluginInstalled || loadingLabelsPlugin
+  );
+  //------ Convert the labelsOpsKeys to the same format as the labelsByKeyFromExisingAlerts
+  const labelsByKeyOps = useMemo(() => {
+    return labelsOpsKeys.reduce((acc: Record<string, Set<string>>, label) => {
+      acc[label.name] = new Set();
+      return acc;
+    }, {});
+  }, [labelsOpsKeys]);
+
+  //------- Convert the keys from the ops labels to options for the dropdown
+  const keysFromGopsLabels = useMemo(() => {
+    return mapLabelsToOptions(Object.keys(labelsByKeyOps), labelsInSubform);
+  }, [labelsByKeyOps, labelsInSubform]);
+
+  //------- Convert the keys from the existing alerts to options for the dropdown
+  const keysFromExistingAlerts = useMemo(() => {
+    return mapLabelsToOptions(Object.keys(labelsByKeyFromExisingAlerts), labelsInSubform);
+  }, [labelsByKeyFromExisingAlerts, labelsInSubform]);
+
+  // create two groups of labels, one for ops and one for custom
+  const groupedOptions = [
+    {
+      label: 'From alerts',
+      options: keysFromExistingAlerts,
+      expanded: true,
+    },
+    {
+      label: 'From system',
+      options: keysFromGopsLabels,
+      expanded: true,
+    },
+  ];
+
+  return {
+    loading: loading || isLoadingLabels,
+    labelsByKeyOps,
```

Comment:
I wonder if we could unify the labels in a way the rest of the code doesn't really matter if the label comes from alerts or ops
What if we have a, for example, OriginatedLabel { labelKey: string, origin: '"alerting" | "ops" } 
Then we could do the grouping just before rendering and keep most of the implementation details between alerts and ops in this hook

Reply:
In my last commit I moved the whole logic for creating labels options to the `useCombinedLabels`, so all the implementation details behind how to create these options is in this hook now.